### DISCUSSION TITLE
How can I pass data between sibling components in React?

### DISCUSSION THREAD

**Discussion by KD63799** [community/community#132320] [Other Feature Feedback, Questions, & Ideas] [answered]
How can I pass data between sibling components in React?

**Comment by Thesirix** [comment]
by props this is the onlmy way

**Comment by iamdulanga** [comment]
Hello, @KD63799 

Please describe your problem.
Dulanga.

**Reply by ghost** [reply]
They will never be my family. I do not have one German ethnicity in my family. They need to pay attention to their own family. Spread the message.

**Comment by abdulhanan1000** [comment, marked as answer]
Hello, 
I think this is what you are looking for: lifting the state up to their common parent component.

For example:
Create the Parent Component:
```
import React, { useState } from 'react';
import SiblingA from './SiblingA';
import SiblingB from './SiblingB';

const Parent = () => {
  const [data, setData] = useState('');

  const handleDataChange = (newData) => {
    setData(newData);
  };

  return (
    <div>
      <SiblingA onDataChange={handleDataChange} />
      <SiblingB data={data} />
    </div>
  );
};

export default Parent;
```

Create SiblingA Component:
```
import React from 'react';

const SiblingA = ({ onDataChange }) => {
  const handleChange = (event) => {
    onDataChange(event.target.value);
  };

  return (
    <div>
      <input type="text" onChange={handleChange} />
    </div>
  );
};

export default SiblingA;
```

Create SiblingB Component:
```
import React from 'react';

const SiblingB = ({ data }) => {
  return (
    <div>
      <p>Data from SiblingA: {data}</p>
    </div>
  );
};

export default SiblingB;
```

- SiblingA receives the callback (onDataChange) as a prop and uses it to notify the parent component when the input value changes.
- SiblingB receives the updated data as a prop and displays it.
- The parent component (Parent) holds the state (data) and provides a callback function (handleDataChange) to update this state.

**Reply by ghost** [reply]
No such thing as common parent!

**Comment by rushit-kakadiya** [comment]


**Reply by ghost** [reply]
Cannot pass data between siblings.

**Comment by ghost** [comment]
> How can I pass data between sibling components in React?

Data cannot be passed between anyone not even mother and child.

**Comment by ghostinhershell** [comment]
Thanks for posting in the GitHub Community, @KD63799!

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the New to GitHub category is solely related to conversations around getting started on using GitHub and the community for beginners of GitHub, Community on Discussions, and those new to different areas of the product alike! This question should be in the General category. I've gone ahead and moved it for you. Good luck!

**Comment by wlb-tycoon** [comment]
If your project is large, you can simply do it using `useContext` React hook.
Here, let me explain steps to share data between sibling components.
1. Create a **Context**
    Create a context file to define and export your context.
```
// MyContext.tsx

import React, { createContext, useState, ReactNode } from 'react';

interface MyContextType {
  sharedData: string;
  setSharedData: React.Dispatch<React.SetStateAction<string>>;
}

const MyContext = createContext<MyContextType | undefined>(undefined);

export const MyProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [sharedData, setSharedData] = useState<string>('Initial Data');

  return (
    <MyContext.Provider value={{ sharedData, setSharedData }}>
      {children}
    </MyContext.Provider>
  );
};

export const useMyContext = () => {
  const context = useContext(MyContext);
  if (!context) {
    throw new Error('useMyContext must be used within a MyProvider');
  }
  return context;
};
```
2. Wrap Your Application with the **Provider**
   Wrap your application or a part of your component tree with the `MyProvider` to make the context available to the components.
```
// App.tsx

import React from 'react';
import { MyProvider } from './MyContext';
import SiblingOne from './SiblingOne';
import SiblingTwo from './SiblingTwo';

const App: React.FC = () => {
  return (
    <MyProvider>
      <SiblingOne />
      <SiblingTwo />
    </MyProvider>
  );
};

export default App;
```
3. Access the **Context** in Sibling Components
   Now, you can access the shared context in your sibling components.
   **Sibling One** (to update the shared data):
```
// SiblingOne.tsx

import React from 'react';
import { useMyContext } from './MyContext';

const SiblingOne: React.FC = () => {
  const { setSharedData } = useMyContext();

  const handleChange = () => {
    setSharedData('Data from Sibling One');
  };

  return (
    <div>
      <h2>Sibling One</h2>
      <button onClick={handleChange}>Update Shared Data</button>
    </div>
  );
};

export default SiblingOne;
```
**Sibling Two** (to read the shared data):
```
// SiblingTwo.tsx

import React from 'react';
import { useMyContext } from './MyContext';

const SiblingTwo: React.FC = () => {
  const { sharedData } = useMyContext();

  return (
    <div>
      <h2>Sibling Two</h2>
      <p>Shared Data: {sharedData}</p>
    </div>
  );
};

export default SiblingTwo;
```
**Summary**:
1. _**Create a Context**_: Define a context and a provider component that manages the shared data.
2. _**Wrap Components**_: Use the provider to wrap the components that need access to the context.
3. _**Access the Context**_: Use the `useMyContext` hook in sibling components to read or update the shared data.

This method would allow sibling components to communicate without directly passing props, providing a clean and efficient way to manage shared state.
